:lock: validate contracts instantiated from network address

matthiasgeihs opened the pull request:
Fixes #11 

added contract validation on node setup. needed to change alice.yaml and bob.yaml so that alice, who is run first in the walkthrough, is deploying the contracts and bob is validating them correctly.

sebastianst commented on cmd/demo/node_setup.go:
```diff
@@ -108,6 +108,9 @@ func (n *node) setup() error {
 func (n *node) setupContracts() (err error) {
 	var adjAddr, assAddr common.Address
 
+	// TODO: must ensure deploying node is started first, otherwise contract validation fails
+	// TODO: for better understanding of what happens, log to console when contracts are deployed
```
We try to avoid committing TODO notes. Better create an issue if it's worth it, or fix it before giving into review.

matthiasgeihs replied:
removed the todo notes from the source code, will consider creating issues for them

sebastianst replied:
So for the first TODO, I think that's nothing the impl can handle. The user must know how to use the node properly, so I propose to add this to the README and description of the configuration file.

For the second TODO, the contract addresses are already logged in the last line `	n.log.WithField("Adj", adjAddr).WithField("Asset", assAddr).Debug("Set contracts")`

matthiasgeihs replied:
regarding first todo: the walkthrough already suggests to start alice first and then bob, so this should be ok.

regarding second todo: i still think it would be instructive to have some more information for the walkthrough user on the console. just so he gets an idea of what is going on in our framework. at the moment there is barely any information logged to the console except for what the types himself. anyhow, this can be left for another PR, as it is not closely related to this PR.

sebastianst commented on cmd/demo/node_setup.go:
```diff
@@ -118,7 +121,15 @@ func (n *node) setupContracts() (err error) {
+
+		ctx, cancel := context.WithTimeout(context.Background(), config.Chain.TxTimeout)
+		defer cancel()
+		err = echannel.ValidateAdjudicator(ctx, n.cb, adjAddr)
+		if err != nil {
+			return err
+		}
```
We always use the schema `if err := foo(); err != nil {... }` if we don't need the defined variable(s) outside the `if` block. If we're not lazy we also wrap the error using `errors.WithMessage` (if it already has a stack) or `errors.Wrap` (if it doesn't). It is described very briefly here https://git.perun.network/perun/go-perun/-/wikis/Coding-Code-of-Conduct/General#external-errors-without-a-stack-trace

```suggestion
		if err := echannel.ValidateAdjudicator(ctx, n.cb, adjAddr); err != nil {
			return errors.WithMessage(err, "validating adjudicator contract")
		}
```

matthiasgeihs replied:
suggested changes have been adopted to the codebase

sebastianst commented on cmd/demo/node_setup.go:
```diff
@@ -132,6 +143,13 @@ func (n *node) setupContracts() (err error) {
+
+		ctx, cancel := context.WithTimeout(context.Background(), config.Chain.TxTimeout)
+		defer cancel()
+		err = echannel.ValidateAssetHolderETH(ctx, n.cb, assAddr, adjAddr)
+		if err != nil {
+			return err
+		}
```
```suggestion
		if err := echannel.ValidateAssetHolderETH(ctx, n.cb, assAddr, adjAddr); err != nil {
			return errors.WithMessage(err, "validating asset holder contract")
		}
```
note that the message should describe the "action", usually in gerund form, because it chains nicely when displaying the error like "Error doing foo: accessing bar: validating contract: opening bla: connection refused".

matthiasgeihs replied:
suggested changes have been adopted to the codebase

ggwpez commented on bob.yaml:
```diff
@@ -18,8 +18,8 @@ node:
   url: ws://127.0.0.1:8545
-  adjudicator: deploy
-  assetHolder: deploy
+  adjudicator: 0x94503E14E26A433C0802E04f2aC1Bb1CE77321F5
```
Why did you change these? Same with `alice.yaml`?!

matthiasgeihs replied:
https://github.com/perun-network/perun-eth-demo/pull/17#issue-450349810

walkthrough starts alice first, but bob was the one deploying, and alice was assuming contracts already deployed, so walkthrough failed. hence, i changed the node configs so that alice is deploying and bob is loading from address

ggwpez replied:
I think we should not change the addresses, better just change the `README` to start bob first. `dst-go` already uses the demo to deploy the contracts and uses hard-coded addresses :laughing:  
I also used it in the `prnm` tests so keeping the contract addresses would be good.

matthiasgeihs replied:
i think you should definitely not rely on these contract addresses to be fixed. so these are todo's for the other projects to remove these dependencies. in fact, i think there should be no such dependencies on a demo project after all. the demo must be independent and it must be possible to change the demo at all times, e.g., for demonstrating additional use cases, or for any other purpose you can think of.

anyhow, i can leave the addresses for now and change the readme, but then for the other projects there are still tasks for removing these dependencies.

sebastianst replied:
`alice.yaml` and `bob.yaml` are just example configuration files that are used in the walkthrough example in the `README`. I think it doesn't matter whether it is changed in the yaml files or README, but it needs to be consistent so that anyone can follow the walkthrough and ends up with two working CLI demo clients.

What do you mean @matthiasgeihs by "i think you should definitely not rely on these contract addresses to be fixed."?

sebastianst replied:
After a short clarifying discussion with @matthiasgeihs I agree that it is more intuitive to first start Alice, then Bob and that other projects cannot rely on the yaml files not changing. They are just example files, after all. If another project wants to use the demo cli for, e.g., contract deployment, then they should use their own specific `deployer.yaml` or whatever configuration that the project has control over.

So I also vote for changing the yamls and keep the more intuitive order in the `README` file.

ggwpez replied:
Then we jest rename Alice->Bob and Bob->Alice :raised_hands:   

PS: Sure there "should" be not deps on the demo, but since i know that it deploys the contracts correct, it was easy to use in other projects tests for debugging them (eg. Team Yash).

sebastianst replied:
Sure, but it is more important that the README file is correct and that it is more intuitive to use (first Alice, then Bob).

matthiasgeihs replied:
see latest commit for a resolution

- to keep starting order and contract addresses:
  - swap wallet secret keys in yaml files
  - swap addresses in yaml files

matthiasgeihs committed:
:lock: validate contracts instantiated from network address

- add contract validation in node setup
- print info about contract deployment and validation
- add a note to README that alice needs to be started first
- to keep starting order and contract addresses:
  - swap wallet secret keys in yaml files
  - swap addresses in yaml files

sebastianst reviewed (approved):
LGTM

ggwpez commented:
> LGTM

Check. Merging.
